### PR TITLE
[automodel] Add FSDPv2-compatible context parallelism support.

### PR DESCRIPTION
**Pull request by cspades**
# What does this PR do ?

- Adds context parallelism support for NeMo AutoModel and support for 128K+ context for `nvidia/Llama-3_3-Nemotron-Super-49B-v1`.

**Collection**: `nemo.collections.llm.gpt.model.hf_auto_model_for_causal_lm.py` and `nemo.lightning.pytorch.strategies.fsdp2_strategy.py`

# Changelog

- Added context parallelism from `torch.distributed.tensor.experimental.context_parallel` using [torchtitan](https://github.com/pytorch/torchtitan) as a reference: [Breaking Barriers: Training Long Context LLMs with 1M Sequence Length in PyTorch Using Context Parallel](https://discuss.pytorch.org/t/distributed-w-torchtitan-breaking-barriers-training-long-context-llms-with-1m-sequence-length-in-pytorch-using-context-parallel/215082).
- When context parallelism is used, FSDPv2 will `fully_shard` across the CP device mesh, not the DP device mesh, due to OOM issues when matrixing DP and CP or creating new flattened meshes. (Loss is still reduced across the flattened DP x CP mesh.) So in general, when using CP, maximize CP and only increase DP if you _need_ data parallelism.
- Context parallelism is only supported for SDPA and Flash Attention v2. For more details on how the support is monkey-patched into torch, refer to [this comment](https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/experimental/_attention.py#L1169-L1179).

## Appendix

- The comment, which we also need some way to check on the NeMo side:
```
    # TODO: provide a more robust way to replace SDPA.
    # Currently we use monkey patch to replace scaled_dot_product_attention with the
    # wrapped fn. This is okay if users do `import torch.nn.functional` but will not
    # work if users do `import torch.nn.functional.scaled_dot_product_attention`.
```
- Slack Thread: [NVIDIA-Only](https://nvidia.slack.com/archives/C08GLHTK37U/p1742868077894889)

# Usage

- We strongly recommend using chunked CE (`--use-chunked-ce`), Flash Attention v2 (`--attn-implementation flash_attention_2`), and a large CP size (`--cp-size`) to train on large context.
- For context larger than 64K, we recommend [commenting out this FP32-upcast in Nemotron-49B](https://huggingface.co/nvidia/Llama-3_3-Nemotron-Super-49B-v1/blob/main/modeling_decilm.py#L1185).
- For datasets with short sequence lengths and when training with long context, it's possible for entire (global) batches to have padding / masked tokens or labels (`id: -100`), which leads to `NaN` for the cross-entropy loss.

To run context parallelism on a single Slurm interactive node (8 GPUs):
```
# Llama-1B @ DP1, CP8, 8K Context Length
torchrun --nproc-per-node 8 examples/llm/sft/automodel.py --strategy fsdp2 --num-nodes 1 --devices 8 --dp-size 1 --cp-size 8 --global-batch-size 32 --micro-batch-size 1 --accumulate_grad_batches 32 --lr 3e-6 --seq-length 8192 --max-steps 10000 --log-every-n-steps 1 --limit-val-batches 0.025 --trust-remote-code --attn-implementation flash_attention_2 --use-chunked-ce
```

Same API for multi-node / high CP, use `python` instead of `torchrun`, and just increase all of the parameters and run an `sbatch` script with this:
```
# Slurm MultiNode Parameters
# Required for multi-node. `scontrol` is not installed or accessible in non-Slurm containers deployed on Slurm compute nodes, only usable on the launcher compute node, so it is impossible to run this within an interactive session of a compute node.
export MASTER_ADDR=$(scontrol show hostnames $SLURM_JOB_NODELIST | head -n 1)
export MASTER_PORT=13742    # Random port number, can also get one from $(($RANDOM % 99999)).
export NUM_GPUS=8   # --gpus-per-node on the EOS cluster is fixed to 8, so we are not allowed to specify it, and SLURM_GPUS_PER_NODE is not set. (Other clusters allow you to specify --gpus-per-node.)
export WORLD_SIZE=$(($NUM_GPUS * $SLURM_NNODES))    # WORLD_SIZE is the number of GPUs across all nodes. Computed on the launcher compute node, and shared with all other compute nodes.
```

# Tests

- 1K, DP64, CP1 (No Context Parallelism, Baseline): [meta-llama_Llama-3.2-1B_nodes8_dev8_strat_fsdp2_dp64_cp1_tpNone_seqlen1024_gb64_mb1_lr3e-06_dset_squad](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/3vsxw8l0?nw=nwusercye_nv)
- 16K: [nvidia_Llama-3_3-Nemotron-Super-49B-v1_dev8_strat_fsdp2_dp1_cp32_tpNone_seqlen16384_gb64_mb1](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/sllis0sh/workspace?nw=nwusercye_nv)
- 32K: [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes4_dev8_strat_fsdp2_dp1_cp32_tpNone_seqlen32768_gb64_mb1](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/pbvvq0a1/workspace?nw=nwusercye_nv)
- Comparing CE Impl
  - Non-Chunked CE: [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes4_dev8_strat_fsdp2_dp1_cp32_tpNone_seqlen32768_gb64_mb1_dset_hf_mock](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/9xzudv73?nw=nwusercye_nv)
    - Rebased on main branch to pick up a chunked CE, which might help with an OOM associated with CE!
    - Ignore loss curve... jaggedness, LR was killed off by a rebase.
  - Chunked: [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes4_dev8_strat_fsdp2_dp1_cp32_tpNone_seqlen32768_gb64_mb1_lr0.0003_dset_hf_mock](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/g7js96um?nw=nwusercye_nv)
  - Memory Improvement: 89% -> 69% = 22.5% memory saved!
- 64K: [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes8_dev8_strat_fsdp2_dp1_cp64_tpNone_seqlen65536_gb64_mb1_lr0.0003_dset_hf_mock](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/wkg0bapq/workspace?nw=nwusercye_nv)
  - OOM was happening here: https://huggingface.co/nvidia/Llama-3_3-Nemotron-Super-49B-v1/blob/main/modeling_decilm.py#L1185!
    - We just commented this out: `logits = logits.float()` since it was being upcasted in forward. This needs to be an option in the published Nemotron-49B model, or else we don't scale as well for CP.
  - 70% memory utilization at CP64, holding steady!
- 128K: [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes16_dev8_strat_fsdp2_dp1_cp128_tpNone_seqlen131072_gb64_mb1_lr0.0003_dset_hf_mock](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/20y2mh00?nw=nwusercye_nv)
  - Follows directly from 64K by doubling nodes to 16 nodes / CP128!
- `Squad` 64K: [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes8_dev8_strat_fsdp2_dp1_cp64_tpNone_seqlen65536_gb64_mb1_lr3e-06_dset_squad](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/bsf7lstu/workspace?nw=nwusercye_nv)
  - Requires sequence length `tokenizer.EOS` padding, because the Squad dataset is full of short sequences.
- `Squad` 128K
  - [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes16_dev8_strat_fsdp2_dp1_cp128_tpNone_seqlen131072_gb64_mb1_lr3e-06(_dset_squad)](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/jjm2w9mv?nw=nwusercye_nv)
  - [nvidia_Llama-3_3-Nemotron-Super-49B-v1_nodes16_dev8_strat_fsdp2_dp1_cp128_tpNone_seqlen131072_gb64_mb1_lr3e-06(_dset_squad) (Run #2)](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/8sp21bfy?nw=nwusercye_nv)
  - [meta-llama_Llama-3.2-1B_nodes16_dev8_strat_fsdp2_dp2_cp64_tpNone_seqlen131072_gb64_mb1_lr3e-06](https://wandb.ai/modelopt/nemo-automodel-nemotron-49b-experimental-cye/runs/1a1k48tu?nw=nwusercye_nv)

### CODE REVIEW

**Comment by BoxiangW** [comment]
Can you drop [0b11427](https://github.com/NVIDIA/NeMo/pull/12821/commits/0b11427187e83f258e8f3995825731e949c51446)?

**Review by ko3n1g — APPROVED** [review]

**Review by akoumpa — APPROVED** [review]
I went over the PR, looks great thank you.
